### PR TITLE
[ggj] feat: boxed and primitive value equality in cast expression

### PR DESCRIPTION
**Pull request by xiaozhenliu-gg5**
[1/6] support boxed/primitive type equality in cast expression.

More details can be found in notes: go/javanaut-boxed-primitive-type-equality

### CODE REVIEW

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java [hunk 1]
```diff
@@ -37,13 +38,34 @@
...
-  public static final TypeNode FLOAT = builder().setTypeKind(TypeKind.FLOAT).build();
+  public static final TypeNode BYTE = builder().setTypeKind(TypeKind.BYTE).build();
+  public static final TypeNode CHAR = builder().setTypeKind(TypeKind.CHAR).build();
   public static final TypeNode DOUBLE = builder().setTypeKind(TypeKind.DOUBLE).build();
+  public static final TypeNode FLOAT = builder().setTypeKind(TypeKind.FLOAT).build();
+  public static final TypeNode INT = builder().setTypeKind(TypeKind.INT).build();
   public static final TypeNode LONG = builder().setTypeKind(TypeKind.LONG).build();
+  public static final TypeNode SHORT = builder().setTypeKind(TypeKind.SHORT).build();
+
+  public static final TypeNode BOOLEAN_WRAPPER =
```
[the remark this answers is not in the document]
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/TypeNode.java:49]: Sounds great, done.

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java [hunk 2]
```diff
@@ -52,16 +52,15 @@ public CastExpr build() {
...
             "Numeric and boolean types are not inter-castable");
-      } else {
+      } else if (castType.isPrimitiveType() || exprType.isPrimitiveType()) {
+        // Check that if one type is primitive, and the other is its wrapper class type.
+        // For example: `Integer` and `int` they can be castable.
         Preconditions.checkState(
-            TypeNode.isReferenceType(castExpr.type())
-                && TypeNode.isReferenceType(castExpr.expr().type()),
-            "Reference types can only be casted to reference types");
+            TypeNode.boxedPrimitiveEquality(castType, exprType),
```
[the remark this answers is not in the document]
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java:62]: Cool, thanks! Done.

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java [hunk 3]
```diff
@@ -97,6 +118,15 @@ public static boolean isReferenceType(TypeNode type) {
         && type.reference() != null;
   }
 
+  public static boolean isBoxedTypeEquals(TypeNode type1, TypeNode type2) {
+    if (type2.isPrimitiveType()) {
+      TypeNode type = type1;
+      type1 = type2;
+      type2 = type;
```
Comment (summer-ji-eng) [inline src/main/java/com/google/api/generator/engine/ast/TypeNode.java:125]: Sorry, I am little confused here. Why do we need to do a swap here?

Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/TypeNode.java:125]: the swap ensures that type1 is primitive and type2 is boxed type, then we do not have to worry about the order when passing in the arguments. For example, this method can be used to check `isBoxedTypeEquals(int, Integer)` or `isBoxedTypeEquals(Integer, int)`. Thanks

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java [hunk 4]
```diff
@@ -52,16 +52,15 @@ public CastExpr build() {
...
-      } else {
+      } else if (castType.isPrimitiveType() || exprType.isPrimitiveType()) {
+        // Check that if one type is primitive, and the other is its wrapper class type.
+        // For example: `Integer` and `int` they can be castable.
         Preconditions.checkState(
-            TypeNode.isReferenceType(castExpr.type())
-                && TypeNode.isReferenceType(castExpr.expr().type()),
-            "Reference types can only be casted to reference types");
+            TypeNode.isBoxedTypeEquals(castType, exprType),
+            "Primitive types can only be cast to their corresponding boxed types.");
```
[the remark this answers is not in the document]
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java]: this expression is valid, updated the message to `Primitive types can only be cast to their corresponding boxed types and vice versa.`

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java (hunk 3 above)
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/TypeNode.java]: done.

File: src/test/java/com/google/api/generator/engine/ast/TypeNodeTest.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+package com.google.api.generator.engine.ast;
+
+import static junit.framework.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class TypeNodeTest {
+  @Test
+  public void isBoxedTypeEquals() {
+    assertEquals(TypeNode.isBoxedTypeEquals(TypeNode.INT, TypeNode.INT_OBJECT), true);
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/TypeNodeTest.java:24]: Why not `assertTrue`?

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java [hunk 6]
```diff
@@ -155,4 +187,17 @@ private static TypeNode createPrimitiveArrayType(TypeKind typeKind) {
   private static boolean isPrimitiveType(TypeKind typeKind) {
     return !typeKind.equals(TypeKind.OBJECT);
   }
+
+  private static HashMap<TypeNode, TypeNode> createBoxedTypeMap() {
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/TypeNode.java:191]: We generally return the superinterface when it comes to collections, can we return a Map here instead?

File: src/test/java/com/google/api/generator/engine/ast/CastExprTest.java [hunk 7]
```diff
@@ -73,6 +73,49 @@ public void validCastExpr_basicPrimitiveWidening() {
 
   }
 
+  @Test
+  public void validCastExpr_castPrimitiveToBoxedType() {
+    PrimitiveValue intValue = PrimitiveValue.builder().setValue("3").setType(TypeNode.INT).build();
+    ValueExpr valueExpr = ValueExpr.withValue(intValue);
+    CastExpr expr = CastExpr.builder().setType(TypeNode.INT_OBJECT).setExpr(valueExpr).build();
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/CastExprTest.java:80]: Could we remove `CastExpr expr` since it's unused? Same below.

File: src/test/java/com/google/api/generator/engine/ast/CastExprTest.java [hunk 8]
```diff
@@ -73,6 +73,49 @@ public void validCastExpr_basicPrimitiveWidening() {
...
+  }
+
+  @Test
+  public void validCastExpr_castBoxedTypeToPrimitive() {
+    Variable variable = Variable.builder().setName("x").setType(TypeNode.DOUBLE_OBJECT).build();
+    VariableExpr variableExpr = VariableExpr.builder().setVariable(variable).build();
+    CastExpr expr = CastExpr.builder().setType(TypeNode.DOUBLE).setExpr(variableExpr).build();
+  }
+
+  @Test
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/CastExprTest.java:90]: Great to see that these edge cases were considered. In this case, both `nestedPrimitiveBoxedType` tests are already covered by the tests for casting primitives to objects and vice versa, so we can remove them.

File: src/test/java/com/google/api/generator/engine/writer/JavaWriterVisitorTest.java [hunk 9]
```diff
@@ -727,6 +727,20 @@ public void writeCastExpr_nested() {
     assertEquals(writerVisitor.write(), "((String) ((Object) str))");
   }
 
+  @Test
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/writer/JavaWriterVisitorTest.java:730]: We won't need this test either, since we can assume that CastExpr will behave correctly and there hasn't been a change to `JavaWriterVisitor`.

File: src/test/java/com/google/api/generator/engine/ast/TypeNodeTest.java (hunk 5 above)
Comment (xiaozhenliu-gg5) [reply src/test/java/com/google/api/generator/engine/ast/TypeNodeTest.java:24]: done.

File: src/test/java/com/google/api/generator/engine/ast/CastExprTest.java (hunk 7 above)
Comment (xiaozhenliu-gg5) [reply src/test/java/com/google/api/generator/engine/ast/CastExprTest.java:80]: done.

File: src/test/java/com/google/api/generator/engine/ast/CastExprTest.java (hunk 8 above)
Comment (xiaozhenliu-gg5) [reply src/test/java/com/google/api/generator/engine/ast/CastExprTest.java:90]: done. Thanks!

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java (hunk 6 above)
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/TypeNode.java:191]: noted, done.

File: src/test/java/com/google/api/generator/engine/writer/JavaWriterVisitorTest.java (hunk 9 above)
Comment (xiaozhenliu-gg5) [reply src/test/java/com/google/api/generator/engine/writer/JavaWriterVisitorTest.java:730]: got it, done.

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java [hunk 10]
```diff
@@ -52,11 +52,15 @@ public CastExpr build() {
       // Only numeric types can be casted around.
       if (castType.isPrimitiveType() && exprType.isPrimitiveType()) {
         // Check that the types are either both booleans or both not booleans.
-        // TODO(xiaozhenliu): This is one of the places where we need to check the equality of boxed
-        // and primitive types.
         Preconditions.checkState(
             !(castType.equals(TypeNode.BOOLEAN) ^ exprType.equals(TypeNode.BOOLEAN)),
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/CastExpr.java:56]: Could you also please check if `boolean x = (Boolean) true` and the reverse are valid?

Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java:56]: Yes, verified in local files and unit tests, cast expression below are valid.
```
        boolean flag = (Boolean) true;
        Boolean flag1 = (boolean) true;
```
but since similar test case is covered by `validCastExpr_castBoxedTypeToPrimitive` and `validCastExpr_castPrimitiveToBoxedType`, we won't add extra test for this. Thanks

Comment (miraleung) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java:56]: The test case won't catch this, because the original check looks specifically at the primitive type. Could we please add a test case? 

With the `equals()` change, the logic above won't need to be changed either.

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java [hunk 11]
```diff
@@ -52,11 +52,15 @@ public CastExpr build() {
...
             "Numeric and boolean types are not inter-castable");
+      } else if (castType.isPrimitiveType() || exprType.isPrimitiveType()) {
+        // Check that if one type is primitive, and the other is its wrapper class type.
+        // For example: `Integer` and `int` they can be castable.
+        Preconditions.checkState(
+            TypeNode.isBoxedTypeEquals(castType, exprType),
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/CastExpr.java:62]: As per my comments in the doc, it may be best to alter `TypeNode.equals()` and use that instead of having an explicit `isBoxedTypeEquals` check here.

Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java]: sure, updated the `equals()` method and the condition check. Open to the error message throwing for the second check. Thanks!

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java (hunk 10 above)
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java:56]: I'm a little bit confused here, the test cases would be like below if I'm understanding correctly which are already covered:
```
  @Test
  public void validCastExpr_castPrimitiveToBoxedType() {
    // [Constructing] `(Boolean) true` where `true` is a primitive value that is `boolean` type.
    PrimitiveValue booleanValue = PrimitiveValue.builder().setValue("true").setType(TypeNode.BOOLEAN).build();
    ValueExpr valueExpr = ValueExpr.withValue(booleanValue);
    CastExpr.builder().setType(TypeNode.BOOLEAN_OBJECT).setExpr(valueExpr).build();
  }

  @Test
  public void validCastExpr_castBoxedTypeToPrimitive() {
     // [Constructing] `(boolean) flag` where `flag` is a object value that is `Boolean` type.
    Variable variable = Variable.builder().setName("flag").setType(TypeNode.BOOLEAN_OBJECT).build();
    VariableExpr variableExpr = VariableExpr.builder().setVariable(variable).build();
    CastExpr.builder().setType(TypeNode.BOOLEAN).setExpr(variableExpr).build();
  }
```
are we talking about the whole assignment expression? Thanks

Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java:56]: a draft PR for reference: https://github.com/googleapis/gapic-generator-java/pull/140/files#diff-59533acd4d8a07afff86d7de2cd8606dR106

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java [hunk 12]
```diff
@@ -124,9 +146,10 @@ public boolean equals(Object o) {
     }
 
     TypeNode type = (TypeNode) o;
-    return typeKind().equals(type.typeKind())
-        && (isArray() == type.isArray())
-        && Objects.equals(reference(), type.reference());
+    return isBoxedTypeEquals(type, this)
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/TypeNode.java:149]: Could we please add parentheses around logic groups for clarity? Also, what if one of the boxed types are arrays (which needs a test as well)?

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java (hunk 10 above)
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java:56]: done.

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java (hunk 12 above)
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/TypeNode.java]: of course, good point. Wrapped the strict equality check to `strictEquals()` method. 
`int[]` cannot be cast to `Integer[]` so boxed type array is not equal to its primitive type array. This check is included in `isBoxedTypeEquals()` method, since we the TypeNode we put in the map is without array setting. Added a unit test to verify that. Thanks

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java [hunk 13]
```diff
@@ -52,17 +52,16 @@ public CastExpr build() {
...
         Preconditions.checkState(
             !(castType.equals(TypeNode.BOOLEAN) ^ exprType.equals(TypeNode.BOOLEAN)),
             "Numeric and boolean types are not inter-castable");
       } else {
         Preconditions.checkState(
-            TypeNode.isReferenceType(castExpr.type())
-                && (castExpr.expr().type().equals(TypeNode.NULL)
-                    || TypeNode.isReferenceType(castExpr.expr().type())),
-            "Reference types can only be casted to reference types");
+            castExpr.type().equals(castExpr.expr().type())
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/CastExpr.java:60]: Could we add some parentheses to clarify the order? This currently reads `( cond1 || cond2 && (cond3))`, so the intent of the first three aren't clear.

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java [hunk 14]
```diff
@@ -138,6 +158,24 @@ public int hashCode() {
...
+  private boolean strictEquals(TypeNode other) {
+    return typeKind().equals(other.typeKind())
+        && (isArray() == other.isArray())
+        && Objects.equals(reference(), other.reference());
+  }
+
+  private boolean isBoxedTypeEquals(TypeNode other) {
+    // If both types are primitive/reference type, return false.
+    // Array of boxed/primitive type is not considered equal.
+    if (isPrimitiveType() == other.isPrimitiveType() || isArray() || other.isArray()) {
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/TypeNode.java:170]: We probably don't need the `isPrimitiveType` check, since the map equality check will take care of that.

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java [hunk 15]
```diff
@@ -138,6 +158,24 @@ public int hashCode() {
     return hash;
   }
 
+  private boolean strictEquals(TypeNode other) {
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/TypeNode.java:161]: Consider making this (and `isBoxedTypeEquals`) un-private, and add a `@VisibleForTesting` annotation. That will allow us to add separate tests for clarity.

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java (hunk 14 above)
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/TypeNode.java:170]: Removed the duplicate check and the comments above. thanks for catching this.

File: src/main/java/com/google/api/generator/engine/ast/TypeNode.java (hunk 15 above)
Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/TypeNode.java:161]: sure, sounds reasonable to me, made the two methods as `protected` and `@VisibleForTesting`. This helps me to tested the two methods separately.

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java (hunk 13 above)
Comment (miraleung) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java:60]: Since this is a long condition, could we please put sub-condition into named variables to the overall one is easier to read?

Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java]: Sounds cool, defined a `isReferenceTypeEquals` to check the condition when castType is a reference type.

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java [hunk 16]
```diff
@@ -52,17 +52,16 @@ public CastExpr build() {
       // Only numeric types can be casted around.
       if (castType.isPrimitiveType() && exprType.isPrimitiveType()) {
         // Check that the types are either both booleans or both not booleans.
-        // TODO(xiaozhenliu): This is one of the places where we need to check the equality of boxed
-        // and primitive types.
         Preconditions.checkState(
             !(castType.equals(TypeNode.BOOLEAN) ^ exprType.equals(TypeNode.BOOLEAN)),
             "Numeric and boolean types are not inter-castable");
       } else {
+        boolean isReferenceTypeEquals =
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/CastExpr.java:59]: This logic reads to me as "The cast type is a reference and the expression type is either null or a reference type." Could we please update the name to something that better reflects this? And/or break it down further so we have a variable for the expression type check only?

Also, what is the rationale for the current expression type check instead of `!isPrimitiveType && !equals(TypeNode.VOID)`?

Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java:59]: Broke the condition check further down to `exprType_referenceOrNull` and `validReferenceTypeCast`. ReferenceTypeCast is only valid when castType is referenceType AND exprType is NULL or reference.

`!TypeNode.isPrimitiveType(exprType) && !equals(TypeNode.VOID)` only ensures the TypeKind is `Object`, but misses the check for `exprType.reference()` is not null.

File: src/test/java/com/google/api/generator/engine/ast/CastExprTest.java [hunk 17]
```diff
@@ -84,6 +84,52 @@ public void validCastExpr_basicPrimitiveWidening() {
...
+            .setReference(ConcreteReference.withClazz(Integer.class))
+            .setTypeKind(TypeNode.TypeKind.OBJECT)
+            .build();
+    Variable variable = Variable.builder().setName("intArray").setType(intArray).build();
+    VariableExpr variableExpr = VariableExpr.builder().setVariable(variable).build();
+    assertThrows(
+        IllegalStateException.class,
+        () -> CastExpr.builder().setType(integerArray).setExpr(variableExpr).build());
+  }
+
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/CastExprTest.java:132]: Add a cast to Void type?

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java (hunk 16 above)
Comment (miraleung) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java]: Thanks, this is clearer. `isValidPrimitiveTypeCast` seems to imply that `void` should be disallowed, could we add a test to ensure this is the case?

Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java]: yes, `isValidPrimitiveTypeCast` ensures that void type is not allowed here.
 While we cannot create a variable with `void` type since we have condition check in Variable: `!variable.type().typeKind().equals(TypeNode.TypeKind.VOID), "Variable type cannot be void"`.

File: src/test/java/com/google/api/generator/engine/ast/CastExprTest.java (hunk 17 above)
Comment (xiaozhenliu-gg5) [reply src/test/java/com/google/api/generator/engine/ast/CastExprTest.java:132]: Done.

File: src/test/java/com/google/api/generator/engine/ast/CastExprTest.java [hunk 18]
```diff
@@ -123,4 +169,13 @@ public void invalidCastExpr_castNumericToBoolean() {
         IllegalStateException.class,
         () -> CastExpr.builder().setType(TypeNode.BOOLEAN).setExpr(variableExpr).build());
   }
+
+  @Test
+  public void invalidCastExpr_castToVoidType() {
+    Variable variable = Variable.builder().setName("x").setType(TypeNode.VOID_OBJECT).build();
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/CastExprTest.java:175]: This looks like it tests `(void) x` where `x` is a void variable. Casting an already-void-typed variable will fail, but what about casting an int-typed variable?

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java (hunk 16 above)
Comment (miraleung) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java]: It doesn't though, because this doesn't look like it prevents `int x; (void) x`.

Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java]: Actually we take `void` as a primitive type in `isPrimitiveType()` for TypeNode, so we won't disallow void here. But the assertion above prevents that:
```
      Preconditions.checkState(
          !castType.equals(TypeNode.VOID) && !exprType.equals(TypeNode.VOID),
          "Casting cannot be done on or to void types");
```

File: src/test/java/com/google/api/generator/engine/ast/CastExprTest.java (hunk 18 above)
Comment (xiaozhenliu-gg5) [reply src/test/java/com/google/api/generator/engine/ast/CastExprTest.java:175]: Updated the unit test to `int x =(void) x` . This fails for the first condition check, which does not allow `void` in castType or exprType.

File: src/main/java/com/google/api/generator/engine/ast/CastExpr.java [hunk 19]
```diff
@@ -52,17 +52,18 @@ public CastExpr build() {
...
-        // and primitive types.
         Preconditions.checkState(
             !(castType.equals(TypeNode.BOOLEAN) ^ exprType.equals(TypeNode.BOOLEAN)),
             "Numeric and boolean types are not inter-castable");
       } else {
+        boolean isExprTypeReferenceOrNull =
+            exprType.equals(TypeNode.NULL) || TypeNode.isReferenceType(exprType);
+        boolean isValidReferenceTypeCast =
+            TypeNode.isReferenceType(castType) && isExprTypeReferenceOrNull;
+        boolean isValidPrimitiveTypeCast = castType.equals(exprType);
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/CastExpr.java:63]: Thanks for adding the test. This variable is still confusing to the reader, can we either rename it or add a check that both types are primitive to this condition?

Comment (xiaozhenliu-gg5) [reply src/main/java/com/google/api/generator/engine/ast/CastExpr.java:63]: confused here, the above condition `if (castType.isPrimitiveType() && exprType.isPrimitiveType())` checks the condition when both types are primitive. Here `isValidPrimitiveTypeCast` means when either exprType or castType is primitive, then if they are equal (strictEqual || boxedTypeEqual) then it's a valid cast.

Updated the name `isValidPrimitiveTypeCast ` to `isEqualType` if it helps. 
The whole condition check is:
- if both types are primitive: they should be both boolean, or both not boolean.
- if not: 1. one type is primitive, another one is reference: check boxed/primitive type equality (`isEqualType`)
              2. both reference type: one reference can be cast to null/reference type  (`isValidReferenceTypeCast`)
